docs/docs/tutorials/first-agent/3-persistent-state.md

```diff
@@ -0,0 +1,177 @@
+
+First, we need to import the `MemorySaver` class from LangGraph. Add the import statement to the top of your `chatbot.ts` file:
+
+```ts
+import { MemorySaver } from "@langchain/langgraph";
+```
+Then, update the code that creates the runnable agent to use a checkpointer. As a reminder, it should currently look like this:
+
+```ts
+// Define the graph and compile it into a runnable
+const app = new StateGraph(MessagesAnnotation)
+  .addNode("agent", callModel)
+  .addEdge("__start__", "agent")
+  .addNode("tools", new ToolNode(tools))
+  .addConditionalEdges("agent", shouldUseTool)
+  .addEdge("tools", "agent")
+  .compile();
+```
+
+We need to pass an instance of `MemorySaver` to the `compile` method. Update the last to the following:
+```ts
+.compile({ checkpointer: new MemorySaver() });
+```
+
+This change doesn't affect how the graph runs. All we are doing is saving a checkpoint of the graph state as it works through each node.
+
+## Step 2: Replace manual state track with the checkpointer
+
+Previously, we were manually tracking the state of the conversation using the `messages` array. Now that the graph has a checkpointer, we don't have to track the state manually. 
+
+Let's remove the `messages` array and the code that updates it with messages from the user and agent. Delete the following 3 bits of code from near the bottom of your `chatbot.ts` file:
+
+```ts
+const messages = Array<BaseMessageLike>();
+
+// Add the user's message to the conversation history
+messages.push({ content: answer, type: "user" });
+
+messages.push(output.messages[output.messages.length - 1]);
+```
+
+Since `messages` is no longer defined, we're getting an error now on the following line where the chatbot is invoked:
+
+```ts
+const output = await app.invoke({ messages });
+```
+
+The app still needs us to pass the *new* message from the user when we invoke it, but the checkpointer will save it to the graph's state after that. Update the line to the following:
+
+```ts
+  const output = await app.invoke({
+      messages: [{ content: answer, type: "user" }],
+    },
+    { configurable: { thread_id: "42" } }
+  );
+```
+
+Notice that we are now passing **two** arguments to `invoke()` - the first object contains the messages, and the second object contains the configurable `thread_id`.
+
+We're using the `MessagesAnnotation` helper, which has a reducer that will append the new message to the graph's `messages` state. This way each time we invoke the chatbot it will get the new message and all the previous messages from this conversation thread.
+
+The `Runnable` now has access to a checkpointer to save progress as it executes the graph. To use it, we are providing a `thread_id` value when calling `.invoke()`. In a real application, you'd probably want to generate unique thread IDs using something like UUID or nanoid. For now, we're using a hardcoded value of "42".
```

Comment:
nit: we don't really explain what a `Runnable` is elsewhere, probably best to leave this out? As well as other references. Just "compiled graph"